### PR TITLE
Add fake.ClientSet & fake.MemoryStorage

### PR DESCRIPTION
**Pull request by karlkfi**
- Add fake.ClientSet that builds a fake.Client & fake.DynamicClient with the same backing MemoryStorage. This is useful for tests that use both clients.
- Rewrite fake.Client & fake.DynamicClient to use a backing MemoryStore. This makes the behavior consistent between the two and reduces the cost of adding new apiserver behaviors.
- Add fake.TypedInformerFactory to build informers for fake.Cache
- Add fake.MinimizeUnstructured to minimize unstructed objects for storage. This makes diffing more consistent and less expensive.
- Add fake.SubresourceStorage to handle storing generic sub-resources by delegating to a fake.MemoryStorage
- Add fake.WatchSupervisor and fake.Watcher to help implement fake.MemoryStorage.Watch
- Add kinds.GVKToString & kinds.ObjectSummary to make logging more consistent about references to objects and kinds

Depends on:
- https://github.com/GoogleContainerTools/kpt-config-sync/pull/618
- https://github.com/GoogleContainerTools/kpt-config-sync/pull/622
- https://github.com/GoogleContainerTools/kpt-config-sync/pull/623

### CODE REVIEW

File: pkg/syncer/syncertest/fake/minimize.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+// The object is modified in-place.
+// Maps are modified in-place, but lists may or may not be replaced.
+func MinimizeUnstructured(uObj *unstructured.Unstructured) {
+	if len(uObj.Object) == 0 {
+		return
+	}
+	uObj.Object = minimizeMap(uObj.Object)
+}
+
+func minimizeMap(objMap map[string]interface{}) map[string]interface{} {
```
Comment (sdowell) [inline pkg/syncer/syncertest/fake/minimize.go:32]: Could you add a docstring for these minimize functions? I was able to grasp these after reading the implementation but a docstring would help with faster comprehension

File: pkg/syncer/syncertest/fake/minimize_test.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/cli-utils/pkg/testutil"
+)
+
+func TestMinimizeUnstructured(t *testing.T) {
+	testcases := []struct {
+		name          string
+		input, output *unstructured.Unstructured
+	}{
+		{
```
Comment (sdowell) [inline pkg/syncer/syncertest/fake/minimize_test.go:29]: Looks like a testcase is missing for a nested list interface

File: pkg/syncer/syncertest/fake/minimize.go (hunk 1 above)
Comment (karlkfi) [reply pkg/syncer/syncertest/fake/minimize.go:36]: done.

File: pkg/syncer/syncertest/fake/minimize_test.go (hunk 2 above)
Comment (karlkfi) [reply pkg/syncer/syncertest/fake/minimize_test.go:29]: added

**Review by sdowell — APPROVED** [review]

**Pull request merged**